### DISCUSSION TITLE
Is arm64-v8a supported?

### DISCUSSION THREAD

**Discussion by ninomsu** [open-telemetry/opentelemetry-cpp#2185] [Q&A]
**CMake Commands:** 

- sudo cmake .. -DBUILD_TESTING=OFF -DWITH_OTLP=ON -DWITH_OTLP_HTTP=ON -DWITH_OTLP_GRPC=OFF -DWITH_LOGS_PREVIEW=OFF -DCMAKE_CXX_STANDARD=17 -DWITH_EXAMPLES=OFF -DOPENTELEMETRY_INSTALL=ON -DCMAKE_BUILD_TYPE=Release -DANDROID_ABI="arm64-v8a" -DANDROID_NATIVE_API_LEVEL=31 -DCMAKE_TOOLCHAIN_FILE=/home/ndk/build/cmake/android.toolchain.cmake

- sudo cmake --build . --target all
 

**Response:** 

**I get many errors similar to:**

opentelemetry-cpp/api/include/opentelemetry/nostd/./internal/absl/types/../utility/../base/internal/../../meta/type_traits.h:84:27: error: reference to 'is_trivially_destructible' is ambiguous
                    absl::is_trivially_destructible<T>::value> {};
                          ^
/usr/local/include/absl/meta/type_traits.h:208:12: note: candidate found by name lookup is 'absl::is_trivially_destructible'
using std::is_trivially_destructible;

**I found the following warning in that file (type_traits.h):**

// WARNING: use of many of the constructs in this header will count as "complex
// template metaprogramming", so before proceeding, please carefully consider
// https://google.github.io/styleguide/cppguide.html#Template_metaprogramming
//
// WARNING: using template metaprogramming to detect or depend on API
// features is brittle and not guaranteed. Neither the standard library nor
// Abseil provides any guarantee that APIs are stable in the face of template
// metaprogramming. Use with caution.

**I'm not too familiar with meta-programming so I'm confused on how to proceed.**

**Comment by owent** [comment]
You may try use `-DWITH_ABSL` if there is a installed abseil-cpp in your system. Or try `-DCMAKE_FIND_ROOT_PATH_MODE_LIBRARY=ONLY`, `-DCMAKE_FIND_ROOT_PATH_MODE_INCLUDE=ONLY`, `-DCMAKE_FIND_ROOT_PATH_MODE_PACKAGE=ONLY ` to tell cmake not find packages from system PATHs.

> For personal use, I utilize https://github.com/atframework/cmake-toolset to build otel-cpp and its dependencies on various platforms and architectures, including arm64.

**Reply by ninomsu** [reply]
Hi, thanks for responding!

What does your environment look like when compiling?

Specifically:

What toolchain are you using?
Which version of absl are you using?
Which version of protobuf are you using?
Did you compile protobuf libraries for arm64 as well? Did you face any linking issues with these arm64 libraries and source files generated from protoc for otlp http?

**Reply by owent** [reply]
We use https://github.com/atframework/cmake-toolset on mac and windows to build arm64 libraries of iOS and Android.

> What toolchain are you using?

https://github.com/atframework/cmake-toolset 

> Which version of absl are you using?

We use several versions of absl, from 20200225 to 20230125.

> Which version of protobuf are you using?

We use several versions of absl, from 3.13 to 3.21 are tested.

> Did you compile protobuf libraries for arm64 as well?

Yes

> Did you face any linking issues with these arm64 libraries and source files generated from protoc for otlp http?

Yes, we found some linking problems when using protobuf 3.20 and 3.21, and we use patch files (https://github.com/atframework/cmake-toolset/blob/main/ports/protobuf/protobuf-v3.20.1.patch and https://github.com/atframework/cmake-toolset/blob/main/ports/protobuf/protobuf-v3.21.4.patch) to solve these problems.
